ydb/library/yql/dq/actors/spilling/spiller_memory_reporter.cpp

```diff
@@ -0,0 +1,48 @@
+#include "spiller_memory_reporter.h"
+
+namespace NYql::NDq {
+
+TSpillerMemoryUsageReporter::TSpillerMemoryUsageReporter(
+    TMemoryReportCallback reportAllocCallback,
+    TMemoryReportCallback reportFreeCallback)
+    : ReportAllocCallback_(reportAllocCallback)
+    , ReportFreeCallback_(reportFreeCallback)
+{
+    Y_ENSURE(reportAllocCallback != nullptr);
+    Y_ENSURE(reportFreeCallback != nullptr);
+}
+
+void TSpillerMemoryUsageReporter::ReportAlloc(ui64 bytes) {
+    ReportAllocCallback_(bytes);
+    BytesAllocated_ += bytes;
+}
+
+void TSpillerMemoryUsageReporter::ReportFree(ui64 bytes) {
+    // TODO(mfilitov): generate error if bytes > BytesAllocates_
+    ui64 toFree = std::min(BytesAllocated_, bytes);
+    if (toFree) {
+        ReportFreeCallback_(toFree);
+    }
+    BytesAllocated_ -= toFree;
+}
+
+TSpillerMemoryUsageReporter::~TSpillerMemoryUsageReporter() {
+    // TODO(mfilitov): generate error if BytesAllocated_ != 0
```

Comment:
If it's somehow legal, then it's better to describe behavior in the comments - why cancelled request with spilling doesn't report freed memory?

Reply:
As I mentioned in the other comment, spilling reports memory as Allocated when it packs values, and reports it as Freed once it gets confirmation that the blob was written to disk.

In this case, if the query is cancelled for any reason between these two events, the memory release won’t be reported. Even though the memory itself will actually be freed.